Add keyword hybridpair for compute_pressure

jdevemy opened the pull request:
Add the keyword hybridpair in compute_pressure

**Summary**

_Add the possibility to use compute_pressure only on a pair type when using `hybrid` or `hybrid/overlay` with multiple pair types_

**Author(s)**

_Alain Dequidt_
_Julien Devemy_

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

mkanski commented:
This looks interesting, but some things are missing:
1. You should update the docs for `compute pressure`.
1. There should be a check if `hybridpairflag-1` isn't larger than number of styles in `pair_style hybrid`. This check is needed, if there are at least two keywords used.
1. Specifying `hybridpairflag` lower than zero should lead to an error.
1. The compute should check if `pair_style hybrid` is actually used and not just silently ignore the keyword.


It would be nice to have an option to compute pressure from more than one substyle at once. The support for `hybrid/overlay` would be also appreciated.

akohlmey commented:
I would recommend against indexing the substyles, but rather use the `force->pair_match()` method, as it is used in other places in LAMMPS, e.g. in `compute pair` or `fix adapt` (which use similar, but slightly different command line conventions)

jdevemy commented:
Thanks for the comments. This request need some modifications, I agree.

But we discovered another problem: with the current pull request code, the `compute_pressure hybridpair` always outputs 0.
It's because the virial is computed at the pair_hybrid level and not at the subpair level.

We manage to correct this by manually adding the flag `no_virial_fdotr_compute = 1` in pair_hybrid.cpp but this is not a clean way to handle this issue.

In your opinion what would be the cleanest way to deal with this issue ?

akohlmey commented:
> Thanks for the comments. This request need some modifications, I agree.
> 
> But we discovered another problem: with the current pull request code, the `compute_pressure hybridpair` always outputs 0.
> It's because the virial is computed at the pair_hybrid level and not at the subpair level.
> 
> We manage to correct this by manually adding the flag `no_virial_fdotr_compute = 1` in pair_hybrid.cpp but this is not a clean way to handle this issue.
> 
> In your opinion what would be the cleanest way to deal with this issue ?

you would not hardcode this into pair style hybrid, but into compute pressure. you can check in its `init()` function if `pair != NULL` and then set `pair->no_virial_fdor_compute = 1`. I've always been thinking, that it might be useful to have this kind of feature as a `pair_modify` option (for testing/debugging). The trick is, that you may set this flag to 1, but never to 0, as some pair styles won't work with fdotr.

jdevemy commented:
I've updated the doc and modified the `compute_pressure` to use `force->pair_match()` and to set the flag `pair->no_virial_fdor_compute = 1`.

Tell-me if it's OK for you.

Regards !

akohlmey commented:
> I've updated the doc and modified the `compute_pressure` to use `force->pair_match()` and to set the flag `pair->no_virial_fdor_compute = 1`.
> 
> Tell-me if it's OK for you.
> 
> Regards !

you must have missed something. all integration tests are failing to compile your PR branch now.

jdevemy commented:
> you must have missed something. all integration tests are failing to compile your PR branch now.

Oups... I've forgot to add the modified compute_pressure.h. Corrected !

sjplimp commented on doc/src/compute_pressure.txt:
```diff
@@ -16,12 +16,13 @@ ID, group-ID are documented in "compute"_compute.html command
 compute 1 all pressure thermo_temp
-compute 1 all pressure NULL pair bond :pre
+compute 1 all pressure NULL pair bond
+compute 1 all pressure NULL hybridpair lj/cut :pre
```
more LAMMPS-like name for this keyword would be pair/hybrid, i.e. general/specific

sjplimp commented on src/pair_hybrid.h:
```diff
@@ -33,6 +33,7 @@ class PairHybrid : public Pair {
   friend class Respa;
   friend class Info;
   friend class PairDeprecated;
+  friend class ComputePressure;
```
not clear why this is needed, no_virial_fdotr_compute is a public va
raible? may be moot, if my other suggestion is followed

akohlmey replied:
This is a leftover from the original implementation, that didn't use force->pair_match() but would access the `PairHybrid::styles` array, which is protected. It can indeed be removed. Also, the `nsub` class variable is no longer needed because of that change.

sjplimp commented on src/compute_pressure.cpp:
```diff
@@ -152,6 +186,11 @@ void ComputePressure::init()
     nvirial = 0;
+    if (hybridpairflag && force->pair) {
+      PairHybrid *ph = (PairHybrid *) force->pair;
+      ph->no_virial_fdotr_compute = 1;
```
this seems problematic, what if this compute goes away (uncompute), 
then the setting in PH will not be un-done
better way to do this might be, PH in its init_style() sets this no_virial
flag in its usual way (currently in flags()), but also loops over computes looking to see
if any (like this one) sets a flag that tells it to override the setting
it would normally make

akohlmey replied:
> this seems problematic, what if this compute goes away (uncompute),
> then the setting in PH will not be un-done

i don't think this would be too drastic, since this use case would be rather unusual.

> better way to do this might be, PH in its init_style() sets this no_virial
> flag in its usual way (currently in flags()), but also loops over computes looking to see
> if any (like this one) sets a flag that tells it to override the setting
> it would normally make

this seems to me like a big hammer for a small problem. after all, the change doesn't break functionality of the affected pair style.

sjplimp replied:
but it makes PH compute something (the virial) inefficiently,
so I think its bad to not remove that requirement if the compute goes away

akohlmey commented:
I just see that as a (small) price to pay for having the feature available.
I worry more about the price we have to pay for maintenance of pair style
hybrid.

Also, the slowdown could be resolved by reissuing the pair style.




--
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

On Mon, Jul 1, 2019, 18:05 Steve Plimpton <notifications@github.com> wrote:

> *@sjplimp* commented on this pull request.
> ------------------------------
>
> In src/compute_pressure.cpp
> <https://github.com/lammps/lammps/pull/1511#discussion_r299234805>:
>
> > @@ -152,6 +186,11 @@ void ComputePressure::init()
>    if (nvirial) {
>      vptr = new double*[nvirial];
>      nvirial = 0;
> +    if (hybridpairflag && force->pair) {
> +      PairHybrid *ph = (PairHybrid *) force->pair;
> +      ph->no_virial_fdotr_compute = 1;
>
> but it makes PH compute something (the virial) inefficiently,
> so I think its bad to not remove that requirement if the compute goes away
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1511?email_source=notifications&email_token=AACGTY53EZ7T4EVI4TK4PEDP5J5RLA5CNFSM4HYIVBD2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5FSNXY#discussion_r299234805>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACGTY23P7UR2WQ5JXTGROLP5J5RLANCNFSM4HYIVBDQ>
> .
>

jdevemy commented:
I've pushed the 2 changes requested from @sjplimp.

sjplimp commented on doc/src/compute_pressure.txt:
```diff
@@ -16,12 +16,13 @@ ID, group-ID are documented in "compute"_compute.html command
 temp-ID = ID of compute that calculates temperature, can be NULL if not needed
 zero or more keywords may be appended
-keyword = {ke} or {pair} or {bond} or {angle} or {dihedral} or {improper} or {kspace} or {fix} or {virial} :ul
+keyword = {ke} or {pair} or {bond} or {angle} or {dihedral} or {improper} or {kspace} or {fix} or {virial} or {hybridpair} :ul
```
also needs pair/hybrid

jdevemy replied:
Sorry. Fixed !

sjplimp commented on doc/src/compute_pressure.txt:
```diff
@@ -67,6 +68,9 @@ extra keywords are listed, then only those components are summed to
 compute temperature or ke and/or the virial.  The {virial} keyword
 means include all terms except the kinetic energy {ke}.
 
+The {hybridpair} keyword means to only include contribution
```
ditto

sjplimp commented:
re: having pair/hybrid set its own no_virial_compute flag correctly at each run by detecting this compute,
that's the best way to do this.  But it's fine to merge this as-is (with the couple of flagged doc page changes).  I can add the logic
later - it's only ~10 lines of code.

akohlmey commented on src/compute_pressure.cpp:
```diff
@@ -140,6 +173,7 @@ void ComputePressure::init()
   nvirial = 0;
   vptr = NULL;
 
+  if (pairhybridflag && force->pair) nvirial++;
```
There is one more thing that should/could be done: it should be checked in the `init()` function, if the pair style is still available. there could have been a pair style command changing it since the compute was defined. please see compute pair for comparison.

jdevemy replied:
Corrected. But I needed to reput `nsub` as in the class.

akohlmey commented:
@sjplimp this has been sitting for a while now waiting for you to approve the changes you requested.
